### ISSUE TITLE
Feature: Notification system

### ISSUE THREAD

**Issue by deepu105** [jhipster/generator-jhipster#1750]
currently JH app generated doesnt have any notification system for below
1. Successfull events like entity creation, update, delete etc.
2. Error events like validation failure on server, unknown errors etc

these can be observed in console logs but to an end user there is no indication of success/failures

My solution: Angular Http interceptor to listen for response, requestError and responseError events and provide a visual notification to user using a toast style notification message(using angular bootstrap alert service). green toast with successful entity creation/update/delete message or red toast with error reason and message

I have implemented this for my app looks good and works well. So if this would be useful in the default app i would like to add it to JH what say @jdubois @atomfrede @gmarziou @gzsombor @andidev

**Comment by gzsombor** [comment]
I've done similar thing, a bit more, a bit less, my solution is aimed for error reporting, and a nice, internationalized error messages. It is based on a http-interceptor and an 'error-display' directive, with that, you can throw an *CustomParameterizedException("error.missingPriceForPart", part.getName(), part.getSerialNumber())* and in the 'error.json', for "error.missingPriceForPart" : "Missing price for {{params[0]}}, with {{params[1]}} serial number!"

I can contribute this, as well :wink:

**Comment by gmarziou** [comment]
It would be very useful for all API interactions, I also did something similar with home made toast style notifications only for sucesses, for failures it depended on context.

For other kind of errors, maybe we could use [$exceptionHandler](https://docs.angularjs.org/api/ng/service/$exceptionHandler), the default implementation simply logs the error into the browser console, we could have custom one that would do more maybe depending on the spring profile

 and we could have a different behavior in dev like showing more details if any

**Comment by deepu105** [comment]
@gzombor @gmarziou thats awesome we should merge our solutions to become a great error handling and notification system, ill commit my fork and let you guys know take a look and see how we can put them all together 
@jdubois what do u think

**Comment by deepu105** [comment]
@gzsombor @gmarziou plz take a look at my commit here https://github.com/deepu105/generator-jhipster/tree/notification

let me know what you guys think.

note: There is a weird issue, the close button in the toast is hidden by a ng-hide class, i have no idea why, it works in my app. so that needs fixing

**Comment by deepu105** [comment]
https://github.com/deepu105/generator-jhipster/commit/98c0b809ef209751b2c10a65d19902cf28b9a2a3

**Comment by gzsombor** [comment]
Here is my implementation : https://github.com/gzsombor/generator-jhipster/commit/051d0639372b674cb30687c712971345423018b5

**Comment by deepu105** [comment]
@gzsombor looks good, it will be great if we can merge these

**Comment by gzsombor** [comment]
@deepu105 : Seems nice, the only thing, which I found annoying in your solution is that it will display the alert for POST even if it's not related to entity creation. I would rather put a configuration parameter inside the resource definition, to enable the alerts (I do have a couple of POST for state changes, complex searches, where the user will be seriously surprised, if that alert is presented :wink:  )

**Comment by deepu105** [comment]
Oh ya i didn't think of that one, coz all my posts were doing creation. I guess yes that can be improved

**Comment by Qinwroks** [comment]
when a PR will be submit? this can be a real key feature!

**Comment by gzsombor** [comment]
I going to merge some code from @deepu105 -s branch, and after we could add it to JHipster. Or I'm open to more ideas, what should it include too :smile:

**Comment by deepu105** [comment]
@gzsombor lets add the basic feature from our forks and then improve on top of it

**Comment by gzsombor** [comment]
Here it goes !

**Comment by deepu105** [comment]
@gzsombor looks good to start with. @jdubois can you plz merge this if you
are ok with it

**Comment by jdubois** [comment]
OK that's a lot of work and a lot of new code.
That reminds me that for JHipster 3.0 we should create our own JHipster library (a JAR file that would be on Maven central): there is no need to generate all that code for users, they will not modify it.
-> I have no issue with the JS code, but for the Java part we add a new "api" package, add some DTOs to handle errors, etc, and that's quite a lot of code, that shouldn't be generated in the end

**Comment by deepu105** [comment]
@jdubois I guess you are right, we can identify stuff that people will not modify usually and put them up in a jar.

Btw the angular alert service introduced can be used by JH users throughout the application to put up toast style notifications like `AlertService.success("Message")` where success can be replaced with `error`, `info` or `warning` we can add it to the docs i guess

**Comment by jdubois** [comment]
@deepu105 oh yes, all this needs to be documented, this is a huge part I forgot!!!

**Comment by jdubois** [comment]
As this is merged I'm closing it, @deepu105 can you write some documentation on the website?

**Comment by deepu105** [comment]
@jdubois yup sure

**Comment by deepu105** [comment]
#1775 is merged
@gzsombor is it possible to move all the error i18n content to error.json instead of global.json?  I couldn't figure out where it is setup to use global.json

**Comment by deepu105** [comment]
@gzsombor else we can have it in the entity json as well so that during entity generation we can generate the i18n keys for the field names

**Comment by gzsombor** [comment]
Yes, we can move it to the error.json, but generally it's not a good idea, to have lot's of localization files - it makes the initialization more complex (That error.json would be needed on every page, so it would be loaded, when the global.json loaded) - and having lot's of http requests could cause other problems as well. For example, on mobile / slow connection, it could even fail with timeouts, so it could cause broken screens.

**Comment by deepu105** [comment]
But in this case it need not be loaded on every page right?  As the directive is init only once in the index.html wont it be sufficient to load error.json once during initial load?

**Comment by gzsombor** [comment]
Yes, but what the reason for splitting the error messages from global.json to error.json, if it's needs to be loaded always together, in the very beginning of the app lifecycle ? You are wasting your resources. 
 Ideally, JHipster should have 3-4 separate json files for the built-in functionalities, (for example 1 for the anonym usage, 1 for the logged in user, and 1 for the admin screens). The other localizations - the entity.jsons and the enum.jsons should be merged into module.json. Of course, I know, it could be a lot of  work, for not too much benefits to introduce modules on the client side, but that would be nice :)

**Comment by deepu105** [comment]
@gzsombor yes i get your point and yes that is better for performance and resource usage but imagine if you have 100s of entities and add all the fields in a single file for translation is gonna be very tough to manage. But i dont have a solution for this as im not very familiar with angular translate. My original consideration of error.json was just becoz it is already present and user shouldnt be confused as we normally start the messages with the filename as prefix except for global i guess

**Comment by gmarziou** [comment]
You could have best of both worlds by having separate files for ease of maintenance combined into one at build time like we do for javascript files.

**Comment by deepu105** [comment]
@gmarziou  ya thats an option. Does grunt/gulp uglify task work for json files?

**Comment by gmarziou** [comment]
I don't know but there are dedicated tasks to minify json and to concatenate them.

**Comment by deepu105** [comment]
Might be worth considering, but will it affect how translate services load
files when the routes are loaded?

On Sat, Aug 1, 2015 at 6:15 PM, Gaël Marziou <notifications@github.com>
wrote:

> I don't know but there are dedicated tasks to minify json and to
> concatenate them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jhipster/generator-jhipster/issues/1750#issuecomment-126895095>
> .
>



-- 
thank you,
Deepu.k.s

**Comment by gmarziou** [comment]
It depends on how we decide to do it, I'm not yet clear in how works the notification system, I need to play with it. Anyway, I think minification of i18n files could make sense anyway for a prod build and I'm even wondering if they are versioned versus caching. Overall, I think we should treat them as style sheets having fragments minifiied and merged into one file, that would probably simplify some of our client code.
